src/libDirectoryService/FinalBlockPreProcessing.cpp

```diff
@@ -573,32 +585,189 @@ bool DirectoryService::CheckMicroBlockHashes()
+
+        m_consensusObject->SetConsensusErrorCode(
+            ConsensusCommon::FINALBLOCK_MISSING_HASH);
+
+        return false;
+    }
+
+    return true;
+}
+
+bool DirectoryService::OnNodeMissingMicroBlocks(
+    const std::vector<unsigned char>& errorMsg, unsigned int offset,
+    const Peer& from)
+{
+    LOG_MARKER();
+
+    if (errorMsg.size() < sizeof(uint32_t) + sizeof(uint64_t) + offset)
+    {
+        LOG_GENERAL(WARNING, "Malformed Message");
+        return false;
+    }
+
+    uint32_t numOfAbsentHashes
+        = Serializable::GetNumber<uint32_t>(errorMsg, offset, sizeof(uint32_t));
+    offset += sizeof(uint32_t);
+
+    uint64_t blockNum
+        = Serializable::GetNumber<uint64_t>(errorMsg, offset, sizeof(uint64_t));
+    offset += sizeof(uint64_t);
+
+    vector<MicroBlockHashSet> missingMicroBlockHashes;
+
+    for (uint32_t i = 0; i < numOfAbsentHashes; i++)
+    {
+        MicroBlockHashSet mbHash;
+        if (mbHash.Deserialize(errorMsg, offset) != 0)
+        {
+            LOG_GENERAL(WARNING, "Unable to deserialize MicroBlockHashSet");
+            return false;
+        }
+        offset += mbHash.size();
+
+        missingMicroBlockHashes.emplace_back(mbHash);
+    }
+
+    uint32_t portNo
+        = Serializable::GetNumber<uint32_t>(errorMsg, offset, sizeof(uint32_t));
+
+    uint128_t ipAddr = from.m_ipAddress;
+    Peer peer(ipAddr, portNo);
+
+    lock_guard<mutex> g(m_mutexMicroBlocks);
+
+    auto& microBlocks = m_microBlocks[blockNum];
+
+    unsigned int cur_offset = 0;
+    vector<unsigned char> mb_message
+        = {MessageType::DIRECTORY, DSInstructionType::MICROBLOCKSUBMISSION};
+    cur_offset += MessageOffset::BODY;
+    mb_message.push_back(SUBMITMICROBLOCKTYPE::MISSINGMICROBLOCK);
+    cur_offset += MessageOffset::INST;
+    Serializable::SetNumber<uint64_t>(mb_message, cur_offset, blockNum,
+                                      sizeof(uint64_t));
+    cur_offset += sizeof(uint64_t);
+    mb_message.resize(cur_offset + sizeof(uint32_t));
+    unsigned int mb_num_offset = cur_offset;
+    cur_offset += sizeof(uint32_t);
+
+    uint32_t numOfMicroblocksSent = 0;
+
+    for (uint32_t i = 0; i < numOfAbsentHashes; i++)
     {
+        MicroBlock mb;
+
         bool found = false;
-        for (auto& microBlock : m_microBlocks)
+        // O(n^2) might be fine since number of shards is low
+        // If its slow on benchmarking, may be first populate an unordered_set and then std::find
+        for (const auto& microBlock : microBlocks)
         {
             if (microBlock.GetHeader().GetTxRootHash()
-                    == microBlockHash.m_txRootHash
+                    == missingMicroBlockHashes[i].m_txRootHash
                 && microBlock.GetHeader().GetStateDeltaHash()
-                    == microBlockHash.m_stateDeltaHash)
+                    == missingMicroBlockHashes[i].m_stateDeltaHash)
             {
+                mb = microBlock;
```

Comment:
I think we can make this more efficient by doing the serializing here directly so we don't create another temporary microblock copy.

Reply:
Noted and will make all of these optimizing in the next PR